### PR TITLE
Added Go Implementation of SkipLists

### PR DESCRIPTION
**Pull request by FLASH2332**
Skip List Implementation in Go(Solves #30 )

This Go code adds a Skip List, which allows for quick searching, inserting, deleting, and displaying of data. It uses a coin toss method with a probability distribution to assign levels to each node, keeping the list balanced.

- Insertion: Adds a new node at a random level using coin tosses to determine the level, then updates links to maintain the list structure.
  
- Search: Finds a node by moving through different levels and checks if the value exists.

- Deletion: Removes a node by adjusting links to skip over it, and lowers the level if necessary.

- Display: Shows the list's nodes at each level for easy viewing.

Example usage demonstrates creating, modifying, and viewing the Skip List.(Edge cases has been handled)

### CODE REVIEW

File: skipLists.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+func (sl *skipList) insert(value int) {
+	newLevel := coinToss(sl.maxNoOfLevel)
+
+	newNode := &Node{
+		value: value,
+		level: newLevel,
+		next:  make([]*Node, newLevel+1),
+	}
+
```
Comment (FirefoxSRV) [inline skipLists.go:48]: The insert method does not check for duplicates. If duplicates are allowed, the implementation is fine.

File: skipLists.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		}
+		fmt.Println()
+	}
+}
+
+func main() {
+	maxLevels := 10
+	skipList := newSkipList(maxLevels)
+
+	// Test empty list
```
Comment (FirefoxSRV) [inline skipLists.go:190]: Instead of hard-coding maxNoOfLevel, consider dynamically adjusting it based on the number of elements in the skip list. Use the relationship  maxNoOfLevel = log 2 (number of elements)

**Review by FirefoxSRV — CHANGES_REQUESTED** [review]
Edge case incorporation

File: skipLists.go (hunk 1 above)
Comment (FLASH2332) [reply skipLists.go:48]: // Test duplicate insertions
	skipList.insert(30)
	skipList.insert(30)
	skipList.display()
This test case ensured that duplicate elements are inserted and displayed.
Output:
Skip List:
Level 9:
Level 8:
Level 7:
Level 6:
Level 5:
Level 4:
Level 3:
Level 2:
Level 1: 30
Level 0: 30 30

File: skipLists.go (hunk 2 above)
Comment (FLASH2332) [reply skipLists.go]: > Instead of hard-coding maxNoOfLevel, consider dynamically adjusting it based on the number of elements in the skip list. Use the relationship maxNoOfLevel = log 2 (number of elements)

The `maxNoOfLevel` is now dynamically adjusted based on the relationship `maxNoOfLevel = log2(number of elements)`, ensuring efficient use of levels as the skip list grows or shrinks.

**Comment by FLASH2332** [comment]
> Edge case incorporation

The changes has been made and tested with few cases.

File: skipLists.go (hunk 1 above)
Comment (FirefoxSRV) [reply skipLists.go:55]: Cool then

**Review by FirefoxSRV — APPROVED** [review]

File: skipLists.go (hunk 2 above)
Comment (FirefoxSRV) [reply skipLists.go]: Yep, good job

**Comment by FLASH2332** [comment]
I was not sure about the location of the file should I make any changes for that?

File: skipLists.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+import (
+	"fmt"
+	"math"
+	"math/rand"
+)
+
+type Node struct {
+	value int
+	next  []*Node // This allows each node in the skip list to have multiple pointers to other nodes
+} 				  // at different levels, supporting the layered structure of the skip list.
```
Comment (Ashrockzzz2003) [inline skipLists.go:12]: You can move this above `Node` declaration and mention `next allows`

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]
About the location of this file
`non_linear/skip_list/skip_list.go`

**Comment by FLASH2332** [comment]
The changes has been made and file has been relocated.

**Comment by Ashrockzzz2003** [comment]
> The changes has been made and file has been relocated.

File's missing?
Can you check again

**Comment by FLASH2332** [comment]
> > The changes has been made and file has been relocated.
> 
> File's missing? Can you check again

Apologies, the file has now been added.

File: non_linear/skip_list/skipLists.go
Comment (Ashrockzzz2003) [inline non_linear/skip_list/skipLists.go:1]: Rename file to `skip_list.go`. All files across repo is named this way.

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]

**Review by Ashrockzzz2003 — APPROVED** [review]
LGTM
